x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/BranchingStep.java

```diff
@@ -0,0 +1,109 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.indexlifecycle;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.index.Index;
+
+import java.util.Objects;
+import java.util.function.BiPredicate;
+
+public class BranchingStep extends ClusterStateActionStep {
+    public static final String NAME = "branch";
+
+    private static final Logger logger = LogManager.getLogger(BranchingStep.class);
+
+    private StepKey nextStepKeyOnFalse;
+    private StepKey nextStepKeyOnTrue;
+    private BiPredicate<Index, ClusterState> predicate;
+    private SetOnce<Boolean> predicateValue;
+
+    /**
+     * {@link BranchingStep} is a step whose next step is based on
+     * the return value of a specific predicate.
+     *
+     * @param key                the step's key
+     * @param nextStepKeyOnFalse the key of the step to run if predicate returns false
+     * @param nextStepKeyOnTrue  the key of the step to run if predicate returns true
+     * @param predicate          the condition to check when deciding which step to run next
+     */
+    public BranchingStep(StepKey key, StepKey nextStepKeyOnFalse, StepKey nextStepKeyOnTrue, BiPredicate<Index, ClusterState> predicate) {
+        // super.nextStepKey is set to null since it is not used by this step
+        super(key, null);
+        this.nextStepKeyOnFalse = nextStepKeyOnFalse;
+        this.nextStepKeyOnTrue = nextStepKeyOnTrue;
+        this.predicate = predicate;
+        this.predicateValue = new SetOnce<>();
+    }
+
+   @Override
+   public ClusterState performAction(Index index, ClusterState clusterState) {
+       IndexMetaData indexMetaData = clusterState.metaData().index(index);
+       if (indexMetaData == null) {
+           // Index must have been since deleted, ignore it
+           logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), index.getName());
+           return clusterState;
+       }
+       predicateValue.set(predicate.test(index, clusterState));
+       return clusterState;
+   }
+
+    /**
+     * This method returns the next step to execute based on the predicate. If
+     * the predicate returned true, then nextStepKeyOnTrue is the key of the
+     * next step to run, otherwise nextStepKeyOnFalse is.
+     *
+     * throws {@link UnsupportedOperationException} if performAction was not called yet
+     *
+     * @return next step to execute
+     */
+   @Override
+   public final StepKey getNextStepKey() {
+        if (predicateValue.get() == null) {
+            throw new UnsupportedOperationException("Cannot call getNextStepKey before performAction");
```

Comment:
I think this should be `IllegalStateException`, because we've reached a state where we want the next step before the action has been performed

Reply:
++ much better